pkg/peertls/peertls.go

```diff
@@ -34,131 +41,137 @@ var (
 	ErrVerifySignature = errs.Class("tls certificate signature verification error")
 )
 
-// IsNotExist checks that a file or directory does not exist
-func IsNotExist(err error) bool {
-	return os.IsNotExist(err) || ErrNotExist.Has(err)
-}
-
-// TLSFileOptions stores information about a tls certificate and key, and options for use with tls helper functions/methods
-type TLSFileOptions struct {
-	RootCertRelPath string
-	RootCertAbsPath string
-	LeafCertRelPath string
-	LeafCertAbsPath string
-	// NB: Populate absolute paths from relative paths,
-	// 			with respect to pwd via `.EnsureAbsPaths`
-	RootKeyRelPath  string
-	RootKeyAbsPath  string
-	LeafKeyRelPath  string
-	LeafKeyAbsPath  string
-	LeafCertificate *tls.Certificate
-	// Create if cert or key nonexistent
-	Create bool
-	// Overwrite if `create` is true and cert and/or key exist
-	Overwrite bool
-}
+// PeerCertVerificationFunc is the signature for a `*tls.Config{}`'s
+// `VerifyPeerCertificate` function.
+type PeerCertVerificationFunc func([][]byte, [][]*x509.Certificate) error
 
-type ecdsaSignature struct {
-	R, S *big.Int
-}
-
-// VerifyPeerCertificate verifies that the provided raw certificates are valid
-func VerifyPeerCertificate(rawCerts [][]byte, _ [][]*x509.Certificate) error {
-	// Verify parent ID/sig
-	// Verify leaf  ID/sig
-	// Verify leaf signed by parent
-
-	// TODO(bryanchriswhite): see "S/Kademlia extensions - Secure nodeId generation"
-	// (https://www.pivotaltracker.com/story/show/158238535)
-
-	for i, cert := range rawCerts {
-		isValid := false
+func Generate(template, parentTemplate *x509.Certificate, parent, signer *tls.Certificate) (*tls.Certificate, error) {
+	k, err := ecdsa.GenerateKey(authECCurve, rand.Reader)
+	if err != nil {
+		return nil, ErrGenerate.New("failed to generateServerTLS root private key", err)
+	}
 
-		if i < len(rawCerts)-1 {
-			parentCert, err := x509.ParseCertificate(rawCerts[i+1])
-			if err != nil {
-				return ErrVerifyPeerCert.New("unable to parse certificate", err)
-			}
+	if signer == nil {
+		signer = &tls.Certificate{
+			PrivateKey: k,
+			Leaf: &x509.Certificate{
+				PublicKey: &k.PublicKey,
+			},
+		}
+	}
+	sk, ok := signer.Leaf.PublicKey.(*ecdsa.PublicKey)
+	if !ok {
+		return nil, ErrUnsupportedKey.New("%s", reflect.TypeOf(sk))
+	}
 
-			childCert, err := x509.ParseCertificate(cert)
-			if err != nil {
-				return ErrVerifyPeerCert.New("unable to parse certificate", err)
-			}
+	if parent == nil {
+		parent = &tls.Certificate{
+			Certificate: nil,
+		}
+	}
+	if parentTemplate == nil {
+		parentTemplate = template
+	}
+	caCert, err := createCert(
+		template,
+		parentTemplate,
+		parent.Certificate,
+		sk,
+		k,
+		signer.PrivateKey,
+	)
+	if err != nil {
+		return nil, ErrGenerate.Wrap(err)
+	}
```

Comment:
could just do `return createCert(...)` and have createCert return an ErrGenerate wrapped error